doc/tutorial/metaprogramming-arith.md

```diff
@@ -0,0 +1,290 @@
+#print quux
+/-
+def quux : Arith :=
+Arith.Add (Arith.Symbol "x") (Arith.Mul (Arith.Symbol "y") (Arith.Symbol "z"))
+-/
+
+def quuz := fromArith% (("x" :+ "y") :* "z") -- brackets change binding order.
+#print quuz
+/-
+def quuz : Arith :=
+Arith.Mul (Arith.Add (Arith.Symbol "x") (Arith.Symbol "y")) (Arith.Symbol "z")
+-/
+```
+
+
+Writing variables as strings, such as `"x"`  gets old; Wouldn't it be so much
+prettier if we could write `x :* y`, and the macro translated this into `Arith.Mul (Arith.Symbol "x") (Arith.Mul "y")`?
+We can do this, and this will be our first taste of manipulating macro variables using `x.getId`
+instead of just directly evaluating them via `$x`.
+We also write a macro rule for `fromArith%` that translates an identifier into
+a string, using `$(Lean.quote (toString x.getId))`.  (TODO: explain what
+`Lean.quote` does):
+
+```lean,ignore
+-- Have to use french quotes («arith») to refer to `arith` as
+-- `arith` is a keyword after declare_syntax_cat.
+-- We declare identifiers as living in the `arith` syntax category, and
+-- prove an interpretation for them with `fromArith%`
+syntax ident : «arith»  
+
+macro_rules
+  | `(fromArith% $x:ident) => `(Arith.Symbol $(Lean.quote (toString x.getId)))
+```
+
+
+Let's test and see that we can now write expressions such as `x :* y` directly instead of having to write `"x" :* "y"`:
+
+```lean,ignore
+def x_ident := fromArith% x 
+#print x_ident
+/-
+def x_ident : Arith :=
+Arith.Symbol "x"
+-/
+
+def x_plus_y : Arith := fromArith% (x :+ y)
+#print x_plus_y
+/-
+def x_plus_y : Arith :=
+Arith.Add (Arith.Symbol "x") (Arith.Symbol "y")
+-/
+```
+
+We now show an unfortunate consequence of the above definitions. Suppose we want to build `(x :+ y) :+ z)`.
+Since we already have defined `x_plus_y` as `x := y`, perhaps we should reuse it! Let's try:
+
+```lean,ignore
+def x_plus_y_plus_z := x_plus_y :+ z
+#print x_plus_y_plus_z
+/-
+def x_plus_y_plus_z : Arith :=
+Arith.Add (Arith.Symbol "x_plus_y") (Arith.Symbol "z")
+-/
+```
+
+Whoops, that didn't work! What happened? Lean treats `x_plus_y` _itself_ as an identifier! So we need to add some syntax
+to be able to "escape" the `fromArith%` context. Let's use the syntax `<[ $e:term ]>` to mean: evaluate `$e` as a real term,
+not an identifier. The macro looks like follows:
+
+```lean,ignore
+syntax "<[" term "]>" : «arith» -- escape for embedding terms into `Arith`
+macro_rules
+  | `(fromArith% <[ $e:term ]>) => e
+
+```
+
+Let's try our previous example:
+
+```lean,ignore
+def x_plus_y_plus_z2 := fromArith% <[ x_plus_y ]>
+#print x_plus_y_plus_z2
+/-
+def x_plus_y_plus_z2 : Arith :=
+x_plus_y
+-/
+```
+
+Perfect!
+
+In this tutorial, we expanded on the previous tutorial to parse a more
+realistic grammar with multiple levels of precedence, how to parse identifiers directly
+within a macro, and how to provide an escape from within the macro context.
+
+#### Full code listing
+
+```lean,ignore
```

Comment:
I accidentally marked the full listing as ignored as well, but we should have at least those checked by CI

Reply:
Done